### PR TITLE
[SYCL][Graph] Add specification for kernel binary updates

### PR DESCRIPTION
**Pull request by fabiomestre**
Adds the kernel binary update feature to the sycl graph specification. This introduces a new dynamic_command_group class which can be used to update the command-group function of a kernel nodes in graphs.

Implemented in:
* https://github.com/intel/llvm/pull/16154
* https://github.com/intel/llvm/pull/15700

### CODE REVIEW

File: sycl/doc/syclgraph/SYCLGraphUsageGuide.md
Comment (reble) [inline sycl/doc/syclgraph/SYCLGraphUsageGuide.md:508]:
```suggestion
auto myContext = Queue.get_context();
auto myDevice = Queue.get_device();
exp_ext::command_graph Graph{myContext, myDevice};
```

File: sycl/doc/syclgraph/SYCLGraphUsageGuide.md
Comment (reble) [inline sycl/doc/syclgraph/SYCLGraphUsageGuide.md:517]:
```suggestion
      get_kernel_bundle<sycl::bundle_state::executable>(myContext, { myDevice }, builtinKernelIds);
```

File: sycl/doc/syclgraph/SYCLGraphUsageGuide.md
Comment (reble) [inline sycl/doc/syclgraph/SYCLGraphUsageGuide.md:465]:
```suggestion
int *PtrB = malloc_device<int>(1024, Queue)​;
```

**Review by EwanC — APPROVED** [review]

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc
Comment (Bensuo) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc:576]: "Provides" is the language used in the SYCL spec itself, for example at the end of this section: https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:device-class

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc
Comment (Bensuo) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc:574]: Slight concern that this reads like the currently active command group will always be the one that is executed (regardless of whether the graph has been updated) rather than the CGF that was active when the graph was finalized (or updated).

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc
Comment (Bensuo) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc:584]: To me this reads like: you can have CGFs which are not kernel operations in a dynamic command group, but you can't update them. This error actually happens in the constructor for the dynamic command group, so should this read more like:

> Dynamic command-groups can only contain kernel operations. Trying to construct a dynamic command-group with functions that contain other operations will result in an error.

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc
Comment (Bensuo) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc:917]: Could we mention explicitly that updating the range of a node using a dynamic command-group will only update the range of the currently active CGF? It is sort of implied by talking about the fact it won't be reset, but I think it would be good for clarity to specify it explicitly first.

File: sycl/doc/syclgraph/SYCLGraphUsageGuide.md
Comment (Bensuo) [inline sycl/doc/syclgraph/SYCLGraphUsageGuide.md:462]: We should include the using statement in the example code if we're using it. Also the other examples seem to favor `sycl_ext`, would be good to be consistent.

File: sycl/doc/syclgraph/SYCLGraphUsageGuide.md
Comment (Bensuo) [inline sycl/doc/syclgraph/SYCLGraphUsageGuide.md:510]: Same comment about including the using statement and using `sycl_ext`

**Review by Bensuo — APPROVED** [review]
LGTM, just some minor comments about wording.

**Review by hdelan — APPROVED** [review]
LGTM apart from a few questions.

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc
Comment (EwanC) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: We discussed offline that get/set_active_index was the preference, I've pushed a commit to update to that

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc
Comment (EwanC) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc:828]: Discussed offline to stick with dynamic command-group

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc
Comment (EwanC) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: Good shout, mentioned this explicitly

**Comment by EwanC** [comment]
ping @intel/llvm-reviewers-runtime for required approval, also @gmlueck to review spec language

File: sycl/test/abi/sycl_symbols_linux.dump
Comment (againull) [inline sycl/test/abi/sycl_symbols_linux.dump:3029]: dynamic_command_group class was introduced just recently https://github.com/intel/llvm/commit/5c73a20d4df4d9fd6e0e1f1594207c74b8f8d855 and so hasn't been included to any release yet.
Considering that  and that this is experimental extension I believe it is acceptable to break abi in this case.

Tagging @intel/llvm-gatekeepers in case if there is another opinion.

**Review by againull — APPROVED** [review]

**Comment by EwanC** [comment]
@intel/llvm-gatekeepers Can this be merged please

**Comment by sommerlukas** [comment]
> @intel/llvm-gatekeepers Can this be merged please

Merging would associate Fabio's private mail with this commit, which is against policy:
>  This commit will be authored by fabio.m.mestre@gmail.com 

@fabiomestre would need to update their Github profile to use the correct mail address for commits.

**Comment by sommerlukas** [comment]
> > @intel/llvm-gatekeepers Can this be merged please
> 
> Merging would associate Fabio's private mail with this commit, which is against policy:
> 
> > This commit will be authored by [fabio.m.mestre@gmail.com](mailto:fabio.m.mestre@gmail.com)
> 
> @fabiomestre would need to update their Github profile to use the correct mail address for commits.

After further consideration, our policy was mainly to avoid commits using the anonymous `...@user.noreply.github.com` mail addresses, which is not the case here. 

I'm merging this PR.

**Pull request merged**
